### PR TITLE
Import ZIP code from AOSP and use it to build a ZipFileSystem

### PR DESCRIPTION
**Pull request by swankjesse**
This is a start.

### CODE REVIEW

**Comment by swankjesse** [comment]
**The code is not original.** I started with Apache-licensed code imported from AOSP and converted it to Kotlin, then removed things we don’t need for a read-only filesystem API.

Upstream:
https://android.googlesource.com/platform/libcore/+/android-6.0.1_r21/luni/src/main/java/java/util/zip/Zip64.java
https://android.googlesource.com/platform/libcore/+/android-6.0.1_r21/luni/src/main/java/java/util/zip/ZipEntry.java
https://android.googlesource.com/platform/libcore/+/android-6.0.1_r21/luni/src/main/java/java/util/zip/ZipFile.java

**Comment by swankjesse** [comment]
There’s still lots of work to do on this.

**Review by Egorand — APPROVED** [review]

**Review by oldergod — COMMENTED** [review]
🚀

**Comment by yschimke** [comment]
Nice alternative to address issues like https://twitter.com/headius/status/1269077314677035008 https://bugs.openjdk.java.net/browse/JDK-8246714

"URLClassLoader/JAR protocol handler poisons the global JarFile cache under concurrent load"

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipEntry.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+   * `../bar/baz`, for example. If the entry name is being used to construct a filename
+   * or as a path component, it must be validated or sanitized to ensure that files are not
+   * written outside of the intended destination directory.
+   *
+   * At most 0xffff bytes when encoded.
+   */
+  val name: String,
+
+  /**
+   * Returns the comment for this `ZipEntry`, or `null` if there is no comment.
```
Comment (yschimke) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipEntry.kt:44]: the val appears non-null?

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/zip.kt [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+  val entries = mutableMapOf<Path, ZipEntry>()
+  for (i in 0 until record.numEntries) {
+    val newEntry = source.readEntry(
+      isZip64 = zip64EocdRecordOffset != -1L
+    )
+    if (newEntry.localHeaderRelOffset >= record.centralDirOffset) {
+      throw IOException("Local file header offset is after central directory")
+    }
+    val entryName = newEntry.name
+    val path = "/".toPath().div(entryName)
```
Comment (yschimke) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/zip.kt:187]: why .div() here?

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    return when (entry.compressionMethod) {
+      ZipEntry.STORED -> {
+        FixedLengthSource(source, entry.size, truncate = true)
+      }
+      else -> {
+        val inflaterSource = InflaterSource(
+          FixedLengthSource(source, entry.compressedSize, truncate = true),
+          Inflater(true)
+        )
+        FixedLengthSource(inflaterSource, entry.size, truncate = false)
```
Comment (yschimke) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt:98]: I understand about truncated the compressed size so you don't overrun, but why truncate = false here?

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    return entries.keys.mapNotNull { if (it.parent == canonicalDir) it else null }
+  }
+
+  @Throws(IOException::class)
+  override fun source(path: Path): Source {
+    // Make sure this ZipEntry is in this Zip file.  We run it through the name lookup.
+    val canonicalPath = canonicalize(path)
+    val entry = entries[canonicalPath] ?: throw FileNotFoundException("no such file $path")
+    val source = fileSystem.source(zipPath).buffer()
+    val cursor = source.cursor()!!
```
Comment (yschimke) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt:62]: I assume a nicer error to come as a follow up.

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    source.skip((fileNameLength + extraFieldLength).toLong())
+
+    return when (entry.compressionMethod) {
+      ZipEntry.STORED -> {
+        FixedLengthSource(source, entry.size, truncate = true)
+      }
+      else -> {
+        val inflaterSource = InflaterSource(
+          FixedLengthSource(source, entry.compressedSize, truncate = true),
+          Inflater(true)
```
Comment (yschimke) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt:96]: For future - what do some typical workloads look like for processing ZipFiles? Do you plan to later reset and reuse Inflater instances, or is it cheap enough to create as long as you end() them?

Comment (yschimke) [reply okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt:96]: Thinking about this, what is particularly nice about this file system is that it makes the lifecycle of the ZipFileSystem first class.  Potentially allowing for some closable, or cacheable/poolable, or singleton filesystems.  Compared to the implicit native filesystem or crazy world of JarURLConnection implicit caching etc.

**Comment by yschimke** [comment]
@swankjesse Can I assume the KMM tweaks for this is are in flight? This is going to be great.

**Pull request merged**